node/packages/client/src/cosmos/parser/index.ts

```diff
@@ -42,41 +43,30 @@ export class TransactionParser {
     // messages make best attempt to track where value is transferring to for a variety of tx types
     // logs provide more specific information if needed as more complex tx types are added
     tx.messages.forEach((msg) => {
-      if (msg.from === address) {
-        // send amount
-        const sendValue = new BigNumber(msg.value?.amount ?? 0)
-        if (sendValue.gt(0)) {
-          parsedTx.transfers = aggregateTransfer(
-            parsedTx.transfers,
-            TransferType.Send,
-            this.assetId,
-            msg.from ?? '',
-            msg.to ?? '',
-            sendValue.toString(10)
-          )
-        }
+      // Not all cosmos messages have a value on the message
+      // for example `withdraw_delegator_reward` must look at events for value
+      const value = new BigNumber(msg.value?.amount || valueFromEvents(msg, tx.events) || 0)
+
+      const transferType = msg.from === address ? TransferType.Send : TransferType.Receive
 
+      if (transferType === TransferType.Send) {
         // network fee
         const fees = new BigNumber(tx.fee.amount)
         if (fees.gt(0)) {
           parsedTx.fee = { caip19: this.assetId, value: fees.toString(10) }
         }
       }
 
-      if (msg.to === address) {
-        // receive amount
-        const receiveValue = new BigNumber(msg.value?.amount ?? 0)
-        if (receiveValue.gt(0)) {
-          parsedTx.transfers = aggregateTransfer(
-            parsedTx.transfers,
-            TransferType.Receive,
-            this.assetId,
-            msg.from ?? '',
-            msg.to ?? '',
-            receiveValue.toString(10)
-          )
-        }
-      }
+      parsedTx.transfers = aggregateTransfer(
+        parsedTx.transfers,
+        transferType,
+        this.assetId,
+        msg.from ?? '',
+        msg.to ?? '',
+        value.toString(10),
+        undefined,
+        msg.type
+      )
```

Comment:
Not a blocker by any means and I know this pattern already exists, but I'd love to see this implemented in a more functional way.

I.e. instead of mutating `parsedTx`'s `transfers` and `fees` in the `forEach`, resolve them upfront and use them in `parsedTx`'s creation.